Crash in NodePublishVolume

This happened in a recent run of the CSI E2E tests. Can't really say how to reproduce this, but anyways, this is the log:

```
time="2020-11-25T15:37:22Z" level=debug msg="method called" func="github.com/sirupsen/logrus.(*Entry).Debug" file="/go/pkg/mod/github.com/sirupsen/logrus@v1.4.2/entry.go:277" linstorCSIComponent=driver method=/csi.v1.Node/NodeGetCapabilities nodeID=demo1.linstor-days.at.linbit.com provisioner=linstor.csi.linbit.com req= resp="capabilities:<rpc:<type:GET_VOLUME_STATS > > capabilities:<rpc:<type:EXPAND_VOLUME > > " version=0.10.1
time="2020-11-25T15:37:22Z" level=debug msg="method called" func="github.com/sirupsen/logrus.(*Entry).Debug" file="/go/pkg/mod/github.com/sirupsen/logrus@v1.4.2/entry.go:277" linstorCSIComponent=driver method=/csi.v1.Node/NodeGetCapabilities nodeID=demo1.linstor-days.at.linbit.com provisioner=linstor.csi.linbit.com req= resp="capabilities:<rpc:<type:GET_VOLUME_STATS > > capabilities:<rpc:<type:EXPAND_VOLUME > > " version=0.10.1
time="2020-11-25T15:37:22Z" level=debug msg="looking up resource by CSI volume id" func="github.com/sirupsen/logrus.(*Entry).Debug" file="/go/pkg/mod/github.com/sirupsen/logrus@v1.4.2/entry.go:277" csiVolumeID=pvc-12bcfee8-1cfa-469f-b281-12b125ec9d44 linstorCSIComponent=client
time="2020-11-25T15:37:22Z" level=debug msg="curl -X 'GET' -H 'Accept: application/json' 'http://linstor-cs.linstor.svc:3370/v1/resource-definitions/pvc-12bcfee8-1cfa-469f-b281-12b125ec9d44'"
time="2020-11-25T15:37:22Z" level=debug msg="Status code not within 200 to 400, but 404 (Not Found)\n"
time="2020-11-25T15:37:22Z" level=debug msg="getting assignment info" func="github.com/sirupsen/logrus.(*Entry).Debug" file="/go/pkg/mod/github.com/sirupsen/logrus@v1.4.2/entry.go:277" linstorCSIComponent=client targetNode=demo1.linstor-days.at.linbit.com volume="<nil>"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x12042f2]

goroutine 78 [running]:
github.com/piraeusdatastore/linstor-csi/pkg/client.(*Linstor).GetAssignmentOnNode(0xc00027ce10, 0x16380a0, 0xc000508ae0, 0x0, 0x7ffd1b038bd1, 0x20, 0x0, 0x0, 0x0)
	/buildroot/pkg/client/linstor.go:956 +0x292
github.com/piraeusdatastore/linstor-csi/pkg/driver.Driver.NodePublishVolume(0x1643860, 0xc00027ce10, 0x1646860, 0xc00027ce10, 0x162ae60, 0xc00027ce10, 0x16438c0, 0xc00027ce10, 0x1607e60, 0xc00027ce10, ...)
	/buildroot/pkg/driver/driver.go:309 +0x6d1
github.com/container-storage-interface/spec/lib/go/csi._Node_NodePublishVolume_Handler.func1(0x16380a0, 0xc000508ae0, 0x1405400, 0xc000382480, 0x135759e, 0x1ef, 0xc000589758, 0x75cf38)
	/go/pkg/mod/github.com/container-storage-interface/spec@v1.2.0/lib/go/csi/csi.pb.go:5594 +0x89
github.com/piraeusdatastore/linstor-csi/pkg/driver.Driver.Run.func1(0x16380a0, 0xc000508ae0, 0x1405400, 0xc000382480, 0xc00025c840, 0xc00025c860, 0xc000399a48, 0x592486, 0x13c7aa0, 0xc000508ae0)
	/buildroot/pkg/driver/driver.go:1042 +0x7e
github.com/container-storage-interface/spec/lib/go/csi._Node_NodePublishVolume_Handler(0x146d3a0, 0xc000219140, 0x16380a0, 0xc000508ae0, 0xc00003ad20, 0xc0002a8480, 0x16380a0, 0xc000508ae0, 0xc0005c2480, 0x21c)
	/go/pkg/mod/github.com/container-storage-interface/spec@v1.2.0/lib/go/csi/csi.pb.go:5596 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000298480, 0x1646920, 0xc000092480, 0xc000334200, 0xc00027d410, 0x1e29630, 0x0, 0x0, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:971 +0x486
google.golang.org/grpc.(*Server).handleStream(0xc000298480, 0x1646920, 0xc000092480, 0xc000334200, 0x0)
	/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:1250 +0xcac
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003845f0, 0xc000298480, 0x1646920, 0xc000092480, 0xc000334200)
	/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:690 +0xa5
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/pkg/mod/google.golang.org/grpc@v1.19.0/server.go:688 +0xa5
```

In any case, a 404 should not result in a nil dereference

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Fixed
+- Crash when calling NodePublishVolume on non-existent volume ([#96])
+
+[#96]: https://github.com/piraeusdatastore/linstor-csi/issues/96
+
 ## [0.10.1] - 2020-11-19
 
 ### Changed

--- a/pkg/client/linstor.go
+++ b/pkg/client/linstor.go
@@ -223,9 +223,9 @@ func (s *Linstor) resourceDefinitionToVolume(resDef lapi.ResourceDefinition) (*v
 	return vol, nil
 }
 
-// GetByName retrives a volume.Info that has a name that matches the CSI volume
+// FindByID retrives a volume.Info that has a name that matches the CSI volume
 // Name, not nessesarily the LINSTOR resource name or UUID.
-func (s *Linstor) GetByName(ctx context.Context, name string) (*volume.Info, error) {
+func (s *Linstor) FindByName(ctx context.Context, name string) (*volume.Info, error) {
 	s.log.WithFields(logrus.Fields{
 		"csiVolumeName": name,
 	}).Debug("looking up resource by CSI volume name")
@@ -249,9 +249,9 @@ func (s *Linstor) GetByName(ctx context.Context, name string) (*volume.Info, err
 	return nil, nil
 }
 
-// GetByID retrives a volume.Info that has an id that matches the CSI volume
+// FindByID retrives a volume.Info that has an id that matches the CSI volume
 // id. Matches the LINSTOR resource name.
-func (s *Linstor) GetByID(ctx context.Context, id string) (*volume.Info, error) {
+func (s *Linstor) FindByID(ctx context.Context, id string) (*volume.Info, error) {
 	s.log.WithFields(logrus.Fields{
 		"csiVolumeID": id,
 	}).Debug("looking up resource by CSI volume id")
@@ -486,7 +486,7 @@ func (s *Linstor) CapacityBytes(ctx context.Context, parameters map[string]strin
 // SnapCreate calls linstor to create a new snapshot on the volume indicated by
 // the SourceVolumeId contained in the CSI Snapshot.
 func (s *Linstor) SnapCreate(ctx context.Context, snap *volume.SnapInfo) (*volume.SnapInfo, error) {
-	vol, err := s.GetByID(ctx, snap.CsiSnap.SourceVolumeId)
+	vol, err := s.FindByID(ctx, snap.CsiSnap.SourceVolumeId)
 	if err != nil {
 		return nil, fmt.Errorf("failed to retrieve volume info from id %s", snap.CsiSnap.SourceVolumeId)
 	}
@@ -533,7 +533,7 @@ func (s *Linstor) SnapCreate(ctx context.Context, snap *volume.SnapInfo) (*volum
 
 // SnapDelete calls LINSTOR to delete the snapshot based on the CSI Snapshot ID.
 func (s *Linstor) SnapDelete(ctx context.Context, snap *volume.SnapInfo) error {
-	vol, err := s.GetByID(ctx, snap.CsiSnap.SourceVolumeId)
+	vol, err := s.FindByID(ctx, snap.CsiSnap.SourceVolumeId)
 	if err != nil {
 		return fmt.Errorf("failed to retrieve volume info from id %s", snap.CsiSnap.SourceVolumeId)
 	}

--- a/pkg/client/mock.go
+++ b/pkg/client/mock.go
@@ -47,7 +47,7 @@ func (s *MockStorage) AllocationSizeKiB(requiredBytes, limitBytes int64) (int64,
 	return requiredBytes / 1024, nil
 }
 
-func (s *MockStorage) GetByName(ctx context.Context, name string) (*volume.Info, error) {
+func (s *MockStorage) FindByName(ctx context.Context, name string) (*volume.Info, error) {
 	for _, vol := range s.createdVolumes {
 		if vol.Name == name {
 			return vol, nil
@@ -56,7 +56,7 @@ func (s *MockStorage) GetByName(ctx context.Context, name string) (*volume.Info,
 	return nil, nil
 }
 
-func (s *MockStorage) GetByID(ctx context.Context, id string) (*volume.Info, error) {
+func (s *MockStorage) FindByID(ctx context.Context, id string) (*volume.Info, error) {
 	for _, vol := range s.createdVolumes {
 		if vol.ID == id {
 			return vol, nil
@@ -101,7 +101,7 @@ func (s *MockStorage) SnapCreate(ctx context.Context, snap *volume.SnapInfo) (*v
 		ReadyToUse:     true,
 	}
 
-	vol, _ := s.GetByID(ctx, snap.CsiSnap.SourceVolumeId)
+	vol, _ := s.FindByID(ctx, snap.CsiSnap.SourceVolumeId)
 	vol.Snapshots = append(vol.Snapshots, snap)
 
 	return snap, nil

--- a/pkg/driver/driver.go
+++ b/pkg/driver/driver.go
@@ -301,9 +301,13 @@ func (d Driver) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolum
 	}
 
 	// Retrieve device path from storage backend.
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "NodePublishVolume failed for %s: %v", req.GetVolumeId(), err)
+	}
+
+	if existingVolume == nil {
+		return nil, status.Errorf(codes.NotFound, "NodePublishVolume failed for %s: volume not found", req.GetVolumeId())
 	}
 
 	assignment, err := d.Assignments.GetAssignmentOnNode(ctx, existingVolume, d.nodeID)
@@ -369,7 +373,7 @@ func (d Driver) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeSt
 		return nil, missingAttr("NodeGetVolumeStats", req.GetVolumeId(), "VolumeId")
 	}
 
-	volume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	volume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "NodeGetVolumeStats failed for %s: failed to check if volume exists: %v", req.GetVolumeId(), err)
 	}
@@ -465,7 +469,7 @@ func (d Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) 
 	volumeSize := data.NewKibiByte(data.KiB * data.ByteSize(requiredKiB))
 
 	// Handle case were a volume of the same name is already present.
-	existingVolume, err := d.Storage.GetByName(ctx, req.Name)
+	existingVolume, err := d.Storage.FindByName(ctx, req.Name)
 	if err != nil {
 		return &csi.CreateVolumeResponse{}, status.Errorf(
 			codes.Internal, "CreateVolume failed for %s: %v", req.Name, err)
@@ -521,7 +525,7 @@ func (d Driver) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequest) 
 		return nil, missingAttr("DeleteVolume", req.GetVolumeId(), "VolumeId")
 	}
 
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, err
 	}
@@ -552,7 +556,7 @@ func (d Driver) ControllerPublishVolume(ctx context.Context, req *csi.Controller
 	}
 
 	// Don't try to assign volumes that don't exist.
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal,
 			"ControllerPublishVolume failed for %s: %v", req.GetVolumeId(), err)
@@ -597,7 +601,7 @@ func (d Driver) ControllerUnpublishVolume(ctx context.Context, req *csi.Controll
 		return nil, missingAttr("ControllerUnpublishVolume", req.GetNodeId(), "NodeId")
 	}
 
-	vol, err := d.Storage.GetByID(ctx, req.VolumeId)
+	vol, err := d.Storage.FindByID(ctx, req.VolumeId)
 	if err != nil {
 		return nil, status.Errorf(codes.Internal,
 			"ControllerUnpublishVolume failed for %s: %v", req.GetVolumeId(), err)
@@ -633,7 +637,7 @@ func (d Driver) ValidateVolumeCapabilities(ctx context.Context, req *csi.Validat
 		return nil, missingAttr("ValidateVolumeCapabilities", req.GetVolumeId(), "VolumeCapabilities")
 	}
 
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal,
 			"ValidateVolumeCapabilities failed for %s: %v", req.GetVolumeId(), err)
@@ -880,7 +884,7 @@ func (d Driver) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRequest
 
 		// Handle case where a single volumes snapshots are requested.
 	case req.GetSourceVolumeId() != "":
-		vol, err := d.Storage.GetByID(ctx, req.GetSourceVolumeId())
+		vol, err := d.Storage.FindByID(ctx, req.GetSourceVolumeId())
 		if err != nil {
 			return nil, fmt.Errorf("failed to list snapshots: %v", err)
 		}
@@ -943,7 +947,7 @@ func (d Driver) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeR
 	}
 
 	// Retrieve device path from storage backend.
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "NodeExpandVolume - get resource-definitions %s failed: %v", req.GetVolumeId(), err)
 	}
@@ -973,7 +977,7 @@ func (d Driver) ControllerExpandVolume(ctx context.Context, req *csi.ControllerE
 		return nil, missingAttr("ControllerExpandVolume", req.GetVolumeId(), "VolumeId")
 	}
 
-	existingVolume, err := d.Storage.GetByID(ctx, req.GetVolumeId())
+	existingVolume, err := d.Storage.FindByID(ctx, req.GetVolumeId())
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "ControllerExpandVolume - get resource-definitions %s failed: %v", req.GetVolumeId(), err)
 	}
@@ -1124,7 +1128,7 @@ func (d Driver) createNewVolume(ctx context.Context, req *csi.CreateVolumeReques
 				return &csi.CreateVolumeResponse{}, status.Errorf(codes.NotFound,
 					"CreateVolume failed for %s: snapshot not found in storage backend", req.GetName())
 			}
-			sourceVol, err := d.Storage.GetByID(ctx, snap.CsiSnap.SourceVolumeId)
+			sourceVol, err := d.Storage.FindByID(ctx, snap.CsiSnap.SourceVolumeId)
 			if err != nil {
 				return &csi.CreateVolumeResponse{}, status.Errorf(codes.Internal,
 					"CreateVolume failed for %s: %v", req.GetName(), err)
@@ -1148,7 +1152,7 @@ func (d Driver) createNewVolume(ctx context.Context, req *csi.CreateVolumeReques
 			}
 			logger.Debugf("pre-populate volume from snapshot: %+v", volumeId)
 
-			sourceVol, err := d.Storage.GetByID(ctx, volumeId)
+			sourceVol, err := d.Storage.FindByID(ctx, volumeId)
 			if err != nil {
 				return &csi.CreateVolumeResponse{}, status.Errorf(codes.Internal,
 					"CreateVolume failed for %s: %v", req.GetName(), err)

--- a/pkg/volume/volume.go
+++ b/pkg/volume/volume.go
@@ -484,9 +484,10 @@ type AttacherDettacher interface {
 type Querier interface {
 	// ListAll should return a sorted list of pointers to Info.
 	ListAll(ctx context.Context) ([]*Info, error)
-	GetByName(ctx context.Context, name string) (*Info, error)
-	//GetByID should return nil when volume is not found.
-	GetByID(ctx context.Context, ID string) (*Info, error)
+	// FindByName returns nil when volume is not found.
+	FindByName(ctx context.Context, name string) (*Info, error)
+	// FindByID returns nil when volume is not found.
+	FindByID(ctx context.Context, ID string) (*Info, error)
 	// AllocationSizeKiB returns the number of KiB required to provision required bytes.
 	AllocationSizeKiB(requiredBytes, limitBytes int64) (int64, error)
 	// CapacityBytes determines the capacity of the underlying storage in Bytes.

